lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -149,8 +149,23 @@ class PipelineLowering {
     // Build the clock enable signal: enable && !stall (if applicable)
     Value stageValidAndNotStalled = args.enable;
     Value notStalled;
-    bool hasStall = static_cast<bool>(args.stall);
-    if (hasStall) {
+    bool hasTopLevelStall = static_cast<bool>(args.stall);
+    bool hasStageLevelStall = hasTopLevelStall;
+    if (hasTopLevelStall) {
+      // There is a stall signal - also check whether this is a non-stallable
+      // stage.
+      if (std::optional<ArrayAttr> stallability = pipeline.getStallability()) {
+        bool stageIsStallable =
+            (*stallability)[stageIndex].cast<BoolAttr>().getValue();
+        if (!stageIsStallable) {
+          // This stage is not stallable - the stall signal is ignored.
+          hasStageLevelStall = false;
+        }
+      }
+    }
+
+    if (hasStageLevelStall) {
+      // Stage is confirmed to be stallable - build the clock enable signal.
       notStalled = comb::createOrFoldNot(loc, args.stall, builder);
```

Comment:
There is a tricky issue here.  It would be great if there was a test that caught this.

Say there are 3 stages: s0=stallable, s1=notstallable, s2=stallable

On a given clock cycle, the valid bits of all stages are 1, and stall=1.  Because s1 is not stallable, the token there will advance.  But s2 is stallable, so the token there will not advance.  This that the token at s1 will crash into the token at s2, and one of them will be dropped.

A fix for this is to track (for each stage) record the nearest non-stallable upstream stage.  If a stage is stallable, it must stall if the top-level stall signal is 1, and the valid bit of the nearest non-stallable upstream stage is 0.

This does mean that some "stallable" stages can't really "stall on a dime", and the caller must expect this.  We could encoded this with attributes to be more explicitly (rather than a boolean per-stage stall, there could be an integer attribute which indicates how quickly the stage can react to a stall signal, or something).